Add support for a node selector in the Helm chart.

bmcstdio opened the pull request:
This PR adds support for specifying a node selector for control-plane components in the Helm chart, as requested in #2851.

Closes #2851.

grampelberg commented:
Now that the helm chart is a real installation method, I kinda like leaving this as a helm only feature and not wiring it through to `install`. WDYT @ihcsim @alpeb. Does that work for your use case @bmcstdio?

bmcstdio commented:
@grampelberg I don't really have a use case for this at the moment, I am just new to Linkerd and felt like getting to know the codebase and contributing by looking at open issues. 🙂

grampelberg commented:
Awesome! Thank you for the PR =) We've got a bunch of awesome swag for first time contributors if you're up for filling out the [survey](https://docs.google.com/forms/d/19GLBNfu8XnIETrwEI3ijoWlocAiDp3-j7CW3l4cTnyc/viewform?edit_requested=true).

bmcstdio commented:
Awesome, done - thank you! 🙂

alpeb commented:
I agree with @grampelberg in that we should discourage the addition of options to the CLI unless they cover some obvious and often requested case, and users should instead be directed down the Kustomize or Helm chart road for endless customization options :scroll:

grampelberg commented on charts/linkerd2/templates/controller.yaml:
```diff
@@ -72,6 +72,10 @@ spec:
         {{.ControllerNamespaceLabel}}: {{.Namespace}}
         {{- include "partials.proxy.labels" .Proxy | nindent 8}}
     spec:
+      {{- if .NodeSelector }}
```
It'd be nice to DRY this up by having it be a template function. Check out how `_affinity.tpl` is doing it. Should collapse all of this down to a single call.

bmcstdio replied:
Done!

grampelberg commented on cli/cmd/install.go:
```diff
@@ -419,6 +422,9 @@ func (options *installOptions) recordableFlagSet() *pflag.FlagSet {
 		&options.highAvailability, "ha", options.highAvailability,
 		"Enable HA deployment config for the control plane (default false)",
 	)
+	flags.StringVar(&options.nodeSelectorString, "node-selector", options.nodeSelectorString,
```
I would rather not expose this functionality through `install`.

bmcstdio replied:
Sure, removed!

grampelberg commented on charts/linkerd2/values.yaml:
```diff
@@ -124,3 +124,5 @@ ProxyInjectDisabled: disabled
 ControllerNamespaceLabel: linkerd.io/control-plane-ns
 LinkerdNamespaceLabel: linkerd.io/is-control-plane
+
+NodeSelector: {}
```
Mind adding a comment here to explain what it is and what it does?

bmcstdio replied:
Sure, done!

cpretzer commented:
@bmcstdio have you tried the helm installation process yet?

If you have, then are you okay with adding these options only for helm install, rather than as `install` and `upgrade` options?

bmcstdio commented:
@cpretzer sorry, I didn't have time to test it yet—and haven't had the time to resume work on this either.

bmcstdio commented:
@cpretzer @grampelberg I've made the requested changes, please let me know what your thoughts are! :)

grampelberg commented on charts/linkerd2/templates/_nodeselector.tpl:
```diff
@@ -0,0 +1,8 @@
+{{- if .NodeSelector }}
+{{- toYaml .NodeSelector | trim | nindent 2 }}
+{{- else }}
+  beta.kubernetes.io/os: linux
```
WDYT about just setting the value to this and getting rid of the if/else?

bmcstdio replied:
I agree :+1:

bmcstdio commented:
> For extra credit, adding a test would be amazing.

Could you please guide me through that?

grampelberg commented:
@ihcsim you wouldn't happen to have a pointer to one of the tests?

ihcsim commented:
@grampelberg @bmcstdio The unit tests [are failing](https://github.com/linkerd/linkerd2/pull/3275/checks?check_run_id=221387315) for this PR. If you run `go test ./cli/... --update`, it should automatically update all the test fixtures for you. (Note that `go test ./cli/... --update` will fail the first time, but if you re-run the same command again, the tests will pass.)

btw, the node selector isn't a HA-only option, right?

cpretzer commented:
@bmcstdio thanks for working on this even while you're busy. I'd love to get this pulled in and I'm running the `test` commands that @ihcsim mentioned.

If all goes well, I'll update this ticket and it should be an easy change for you to rebase and commit.

bmcstdio commented:
> btw, the node selector isn't a HA-only option, right?

@ihcsim right, I don't think it is/should be.

---

@cpretzer @grampelberg @ihcsim I've rebased against `master` and fixed the unit tests.

bmcstdio commented:
@alpeb @cpretzer @grampelberg @ihcsim can we get this merged then, please? 🙂

bmcstdio commented:
@alpeb @cpretzer @grampelberg @ihcsim awesome - can we get this merged then, please? 🙂
